jenkins/check_projects.py

```diff
@@ -39,190 +39,414 @@
-  print >>sys.stderr, 'Checking %d projects for iam bindings:' % len(projects)
-  for role, needs in sorted(needed.items()):
-    print >>sys.stderr, '  %s: %s' % (role, ','.join(Sane(n) for n in needs))
-
-  threads = [
-      threading.Thread(target=Check, args=(p, needed, fix))
-      for p in sorted(projects)
-  ]
-
-  for thread in threads:
-    thread.start()
-
-  for thread in threads:
-    thread.join()
-
-  print >>sys.stderr, '%d already configured, %d updated, %d problems' % (
-      len(completed), len(updated), len(errors))
-
-  fixers = collections.defaultdict(int)
-  unk = ['user:unknown']
-  for project in errors:
-    for name in helpers.get(project,unk):
-      fixers[name] += 1
-    print >>sys.stderr, '  %s: %s' % (
-        project, ','.join(Sane(s) for s in sorted(helpers.get(project, unk))))
-
-  print >>sys.stderr, 'Helpers:'
-  for name, count in sorted(fixers.items(), key=lambda i: i[1]):
-    print >>sys.stderr, '  %s: %s' % (count, Sane(name))
-  sys.exit(1)
-
-
-def Sane(member):
-  if ':' not in member:
-    raise ValueError(member)
-  email = member.split(':')[1]
-  return email.split('@')[0]
-
-
-def LoadProjects(configs, filt):
-  projects = set()
-  regex = re.compile(filt)
-  for dirname, _, files in os.walk(configs):
-    for path in files:
-      full_path = os.path.join(dirname, path)
-      if not regex.match(path):
-        continue
-      with open(full_path) as fp:
-        for project in re.findall('^PROJECT="?([^"\r\n]+)"?$', fp.read(), re.MULTILINE):
-          if '{' not in project:
-            projects.add(project)
-            continue
-          else:
-            raise ValueError(project)
-  return projects
-
-
-def Update(project, role, member):
-  with semaphore:
-    err = subprocess.call(
-        [
-            'gcloud', '-q', 'projects',
-            'add-iam-policy-binding',
+class RateLimitedExec(object):
+    """Runs subprocess commands with a rate limit."""
+
+    def __init__(self):
+        self.semaphore = threading.Semaphore(10)  # 10 concurrent gcloud calls
+
+    def check_output(self, *args, **kwargs):
+        """check_output with rate limit"""
+        with self.semaphore:
+            return subprocess.check_output(*args, **kwargs)
+
+    def call(self, *args, **kwargs):
+        """call with rate limit"""
+        with self.semaphore:
+            return subprocess.call(*args, **kwargs)
+
+
+class Results(object):
+    """Results of the check run"""
+
+    class Info(object):
+        """Per-project information"""
+        def __init__(self):
+            self.updated = False
+            # errors is list of strings describing the error context
+            self.errors = []
+            # People with owners rights to update error projects
+            self.helpers = []
+
+    def __init__(self):
+        self.lock = threading.Lock()
+        self.projects = {}  # project -> Info
+
+    @property
+    def errors(self):
+        """
+        Returns:
```

Comment:
I'd recommend reformatting this as `"""Returns set of projects that have errors."""`